### ISSUE TITLE
Improve how "sqlfluff fix" handles parse errors

### ISSUE THREAD

**Issue by barrywhart** [sqlfluff/sqlfluff#2424]
Currently, `sqlfluff fix` applies fixes even if there are parse errors. It merely prints a warning and exits with status code `1`.

We should make it "harder" to apply fixes when there are parse errors, because this can easily cause worse problems (i.e. introducing new errors into the SQL code, because rules often don't handle bad code very well, and this is likely to remain the case).

Some ideas:
* If not in "force" mode, mention there were parse errors when asking whether to apply fixes.
* Should we provide a way to let the user apply fixes to the files that parsed successfully, skipping only the files that had parse errors?
* Should we consider changing "force" mode to a "counter" option, so -f would fail on parse errors without applying fixes, but -ff would go ahead and apply fixes anyway?

Note: This issue is a related follow-on to #2345.

**Comment by tunetheweb** [comment]
I think fix the files with no parsing errors, do not fix files with parsing errors (even if some fixes could potentially be applied to parsable bits of the file) and exit non-zero.

**Comment by jpy-git** [comment]
> because rules often don't handle bad code very well, and this is likely to remain the case

It's not that rules don't handle unparsable code well due to a flaw in their logic or that they can be improved to, it's that unparsable code is a flat structure and therefore does not follow the same logical segment grouping as a normal tree and therefore a rule cannot be expected to interpret that, nor should it. It's the responsibility of the parser to group the segments not a rule, hence unparsable, and if the parser can't interpret the code then the rule shouldn't try to infer logic from it. That's why most linters/static analysis tools take a fail safe and fail early approach to unparsable code.

Note that even if you only disable `fix` for these unparsable segments you would still have `lint` raise the same nonsense lint results that can mislead users into looking for fixes that aren't there and wasting their time.
Take #2306, a good example of one of these corrupting fixes, even if you were to disable the fix in that case, a user could still run `sqlfluff lint` (as many of us do in general use and in pre-commit or CI) and see the nonsense errors that lead to the corrupted fix. They might run `fix` afterwards and wonder why it isn't being fixed and then try and fix it themselves manually even though their statement is fine and SQLFluff just can't parse it.

**Comment by jpy-git** [comment]
> I think fix the files with no parsing errors, do not fix files with parsing errors (even if some fixes could potentially be applied to parsable bits of the file) and exit non-zero.

^ I agree that we shouldn't quit out if only 1 file doesn't parse.
But for my reasoning above, it should be "lint/fix the files with no parsing errors, do not lint/fix files with parsing errors"

**Comment by jpy-git** [comment]
> * Should we consider changing "force" mode to a "counter" option, so -f would fail on parse errors without applying fixes, but -ff would go ahead and apply fixes anyway?

@barrywhart I would advise against providing users with an option to force through these unparsable corrupting fixes for their own safety. There's a lot of risk, especially for less technical users, and IMO the responsible thing is for SQLFluff to accept it cannot fix these issues until the parsing is sorted. `--force` should be reserved simply for skipping the question prompt on whether or not they want to apply the normal fixes.

**Comment by barrywhart** [comment]
My attempt to summarize some of the above:
* `sqlfluff` should _never_ make fixes to file(s) with parse errors. However, it's fine for it to fix other files in the same run that parsed okay. The terminal output should should alert the user to parse errors.
* In order to avoid users getting stuck or frustrated with `pre-commit`, parse errors should not cause `sqlfluff fix` to exit with an error code. For consistency, change `sqlfluff lint` to behave the same way. Rationale: Even with parse errors, the linter often produces useful output. In other words, it's only a _partial_ failure, and let's be optimistic/permissive with the user's workflow, even as we are cautious about making dangerous changes.

This is different from the behavior the user proposed in the original issue (#2345), but I think it's a better overall plan.

**Comment by tunetheweb** [comment]
I think I’ve mentioned this before but I could be persuaded on not fixing code with parsing errors (which was my suggestion above), as agree there is a high degree of risk for a lot of the rules.

Saying that a lot of rules *could still run in fix mode* either completely safely (L001, L002, L004… etc.), or relatively safely (L015, L023…etc.) even with parse errors in other areas of the same SQL code. But it’s not a hill I’m willing to die on, so can accept just blocking fix on unparsable (again, which is why I suggested that, as think it’s the right approach even if over limiting in some cases). We even have a `works_on_unparsable` flag to highlight those rules (though don’t think it’s really used much).

I would, however, be strongly against blocking lint on unparsable. It’s a read-only option and can give a lot of useful other errors even some of the SQL is unparsable. Yes there’s the risk of lint giving out the “wrong information” in this case cause it’s misunderstanding the code but I think that’s 1) relatively unlikely versus the gains from seeing the other errors (in most cases it cannot do anything with that code and correctly skips ahead until it finds code it does recognise as expected) and 2) low risk even if it does that.

This also brings us to the question of `noqa: PRS`. Those are “hidden” or “accepted” unparsable bits of code (either due to a tweak/custom code specific to that team so not worth adding to SQLFluff, or as missing grammar in SQLFluff). That is a valid way of rolling out SQLFluff until grammar is added IMHO (it’s what I did, and why I started contributing to SQLFluff so I could remove them). Removing ability to fix, but especially lint, around those in the meantime is a severe restriction of SQLFluff and I personally would not have rolled it out as fast to my projects (perhaps not at?) if I wasn’t able to use that. And if we accept that `noqa: PRS` shouldn’t bomb out lint, then we’re basically saying unparsable SQL is still useful to lint, so should also attempt to link unparsable SQL even without `noqa: PRS` in my opinion.

**Comment by barrywhart** [comment]
@tunetheweb: Note that "returning an error code from lint" is not the same as "not producing useful output from lint". I'm proposing we leave the lint output as is, only changing the exit code.

I'm open to continuing to tweak this behavior down the road. In this discussion, I've mostly been focused on immediate next steps.

The idea you're proposing (applying some rules but not others) could go nicely with the long-discussed plan of grouping the rules into "core" and other kinds of rules. 💪

**Comment by barrywhart** [comment]
It sounds like it could be useful to make a longer-term plan in this area, defining 2-4 steps or milestones to aim for. By "long term", I'm not necessarily saying a long amount of time, just a direction to go in. Speed and time mostly depends on how busy or motivated we are. 

What do y'all think? What would be a good way to draft and collaborate on that? GitHub wiki? Google doc? Something else? Once we're happy with it, we could create issues and/or add the doc "as is" to the wiki.

**Comment by tunetheweb** [comment]
I think this is one for the next maintainers meeting. Clearly there are some strong opinions here, and us three have had this conversation at least once before, so let’s talk it out to get broad direction before we plan the steps.

**Comment by barrywhart** [comment]
Slack interchange between me and @tunetheweb:

>Me: I’d really like to disable “fix” when there are parse errors. It seems like we have at least 3 issues relating to that.
@tunetheweb: I think that’s a good default. I do think we should maybe include a --fix-even-unparsable (different to --force).

**Comment by barrywhart** [comment]
@tunetheweb also points out:
>Also would it be confusing to users if they --noqa:PRS so now no longer see parsing error in lint but still we won’t fix? Need to be clear on error message.

**Comment by barrywhart** [comment]
Joe Young had provided a [document](https://sqlfluff.slack.com/files/U02NB3FADLN/F02V51G3QRK/sqlfluff_parsing.docx) with some deeper thoughts and rationale on this issue.

**Comment by barrywhart** [comment]
I'm planning to work on this issue.

I think I've found a reasonable middle ground that we can move forward with, without any major objections.

Notes:
* Since we initially discussed this, #2509 enhances `noqa` to support ignoring template (TMP) and parse (PRS) errors. In the context of this issue, we can use that capability to let users "acknowledge" parse errors, thus allowing SQLFluff to report success and not block users from committing (i.e. if they are using SQLFluff with pre-commit).

Proposed behavior:
* `sqlfluff lint`
  * If there's a parse error, print a warning message. Default to "don't lint", **even if the parse error is ignored with "noqa"**. The message should be clear that SQLFluff won't lint even if they ignore parse errors. The only way to lint in the presence of parse errors is to use the new command-line option.
  * If there are still parse errors after applying "noqa" directives, exit 1 (failure). Otherwise, allow exit 0 (success).
  - Add a new lint option, `--LINT-EVEN-UNPARSABLE`: If specified, lints even on parse error.  It's in ALL CAPS to try and make it clear it's a bit "out there". Prints a warning message (in red?) to tell users this option may result in undefined behavior, and that we likely won't fix any issues encountered when using it. 
* `sqlfluff fix`
  - If there's a parse error, print a warning message. Default to "don't lint/fix", **even if the parse error is ignored with "noqa"**.  The message should be clear that SQLFluff won't lint/fix even if they ignore parse errors. The only way to lint/fix in the presence of parse errors is to use the new command-line option.
  - If there are still parse errors after applying "noqa" directives, exit 1. Otherwise, allow exit 0.
  - Add a new fix option, `--FIX-EVEN-UNPARSABLE`: If specified, lints/fixes even on parse error.  It's in ALL CAPS to try and make it clear it's a bit "out there". Prints a warning message (in red?) to tell users this option may result in undefined behavior, and that we likely won't fix any issues encountered when using it.

**Comment by alanmcruickshank** [comment]
Based on the conversation this afternoon I think the behaviour we settled on was:

`sqlfluff lint`:
- Same behaviour as now.

`sqlfluff fix`:
- Only fix if no unparsable segments found in the file. If any are found, no fixes happen even if parsing errors are supressed using noqa or `--ignore` cli options. The only way to override this is with the `--FIX-EVEN-UNPARSABLE` option as above.
- If there are still parse errors after applying "noqa" directives, exit 1. Otherwise, allow exit 0.

**Comment by tunetheweb** [comment]
SGTM

The only other comment we said is we might change `_works_on_unparsable` to apply to whole file (rather than segments as it does currently), which would basically give most of above functionality but potentially allow some “safe” fixes to still be applied?

That could be a follow on from this though if want to do above first. And think we should still exit error on fix (unless that section is `noqa`ed).

**Comment by tunetheweb** [comment]
> ... which would basically give most of above functionality but potentially allow some “safe” fixes to still be applied?

Then again maybe that will be more confusing than it's worth ("why was the file half-fixed?") and so maybe best left to the override?

In which case we should probably document the `_works_on_unparsable` flag to explain it will only be used on `--FIX-EVEN-UNPARSABLE` mode.

**Comment by barrywhart** [comment]
`rules.rst` has the following:
>   It should be noted that ignoring ``TMP`` and ``PRS`` errors can lead to
>   incorrect ``sqlfluff lint`` and ``sqfluff fix`` results as SQLFluff can
>   misinterpret the SQL being analysed.

Review this and other docs as part of this work.

**Comment by barrywhart** [comment]
Test cases:

1
=
SELCT my_col
FROM my_schema.my_table

sqlfluff fix test.sql
Fixes? No.
Exit code: 1

2
=
SELCT my_col -- noqa: PRS
FROM my_schema.my_table

sqlfluff fix test.sql
Fixes? No.
Exit code: 0

2
=
SELCT my_col -- noqa: PRS
FROM my_schema.my_table

sqlfluff fix --FIX-EVEN-UNPARSABLE test.sql
Fixes? Yes.
Exit code: 0

**Comment by tunetheweb** [comment]
Better test case might be:

```sql
SELCT my_col -- noqa: PRS
FROM my_schema.my_table
where 1=2
```

That was `fix` should correct the capitalisation inconsistency of `where`, and ignore the `SELCT` line.

Without that it's difficult to test if `fix` worked in case 3.

**Comment by barrywhart** [comment]
Good point! Just getting my thoughts together for now. This particular task feels like a great fit for TDD. Normally, I write the tests **after** the code (don't tell anyone!).
